cmd/isp/src/protocol.rs

```diff
@@ -0,0 +1,544 @@
+        cnt += r;
+    }
+
+    let frame = FramingPacket::read_from(frame_bytes.as_slice())
+        .ok_or(anyhow!("parse_erorr"))?;
+
+    if frame.header.packet_type != (PacketType::Data as u8) {
+        return Err(anyhow!(
+            "Expected a data packet, got {:x} instead",
+            frame.header.packet_type
+        ));
+    }
+
+    cnt = 0;
+    let length: usize =
+        (frame.length_low as usize) | ((frame.length_high as usize) << 8);
+    let mut response = vec![0; length];
+
+    while cnt != length {
+        let r = port.read(&mut response[cnt..])?;
+        cnt += r;
+    }
+
+    check_crc(&frame_bytes, &response, &frame)?;
+
+    Ok(response)
+}
+
+pub fn read_response(
+    port: &mut dyn serialport::SerialPort,
+    response_type: ResponseCode,
+) -> Result<Vec<u32>> {
+    let mut frame_bytes = vec![0; core::mem::size_of::<FramingPacket>()];
+    let mut cnt = 0;
+
+    while cnt != core::mem::size_of::<FramingPacket>() {
+        let r = port.read(&mut frame_bytes[cnt..])?;
+        cnt += r;
+    }
+
+    let frame = FramingPacket::read_from(frame_bytes.as_slice())
+        .ok_or(anyhow!("Parse_err"))?;
+
+    // A response packet is a specific type of command packet.
+    if frame.header.packet_type != (PacketType::Command as u8) {
+        return Err(anyhow!(
+            "Expected a command, got {:x}",
+            frame.header.packet_type
+        ));
+    }
+
+    cnt = 0;
+    let length: usize =
+        (frame.length_low as usize) | ((frame.length_high as usize) << 8);
+    let mut response = vec![0; length];
+
+    while cnt != length {
+        let r = port.read(&mut response[cnt..])?;
+        cnt += r;
+    }
+
+    check_crc(&frame_bytes, &response, &frame)?;
+
+    let command =
+        RawCommand::read_from(&response[..core::mem::size_of::<RawCommand>()])
+            .ok_or(anyhow!("Parse error"))?;
+
+    if command.tag != (response_type as u8) {
+        return Err(anyhow!(
+            "Expected a response type of {:x}, got {:x}",
+            response_type as u8,
+            command.tag
+        ));
+    }
+
+    let mut params: Vec<u32> = Vec::new();
+    let mut cnt = 0;
+    let mut index = core::mem::size_of::<RawCommand>();
+
+    while cnt < command.parameter_count {
+        params.push(u32::from_le_bytes(response[index..index + 4].try_into()?));
+        cnt += 1;
+        index += 4;
+    }
+
+    send_ack(port)?;
+
+    // First paramter is always the return code;
+    let retval = params[0];
+
+    if retval != 0 {
+        // Some more specific error messages.
+        if retval == 10203 {
+            Err(anyhow!("Did you forget to erase the flash? (err 10203)"))
+        } else if retval == 10101 {
+            Err(anyhow!(
+                "Incorrect signature. Is the SBKEK set correctly? (err 10101)"
+            ))
+        } else {
+            Err(anyhow!("ISP error returned: {}", retval))
+        }
+    } else {
+        Ok(params)
+    }
```

Comment:
This could be a little cleaner as a match statement:
```suggestion
    match params[0] {
        0 => Ok(params),
        10203 => Err(anyhow!("Did you forget to erase the flash? (err 10203)")),
        10101 => Err(anyhow!(
            "Incorrect signature. Is the SBKEK set correctly? (err 10101)"
        )),
        retval => Err(anyhow!("ISP error returned: {}", retval)),
    }
```